cpp/include/cudf/detail/copy_range.cuh

```diff
@@ -0,0 +1,305 @@
+    bool in_range = (index >= begin && index < end);
+
+    // write data
+    if (in_range) data[index] = input.data(input_idx);
+
+    if (has_validity) {  // update bitmask
+      int active_mask = __ballot_sync(0xFFFFFFFF, in_range);
+
+      bool valid = in_range and input.valid(input_idx);
+      int warp_mask = __ballot_sync(active_mask, valid);
+
+      cudf::bitmask_type old_mask = bitmask[mask_idx];
+
+      if (lane_id == 0) {
+        cudf::bitmask_type new_mask = (old_mask & ~active_mask) |
+                                      (warp_mask & active_mask);
+        bitmask[mask_idx] = new_mask;
+        // null_diff = (mask_size - __popc(new_mask)) - (mask_size - __popc(old_mask))
+        warp_null_change[warp_null_change_id] +=
+          __popc(active_mask & old_mask) - __popc(active_mask & new_mask);
+      }
+    }
+
+    input_idx += blockDim.x * gridDim.x;
+    mask_idx += masks_per_grid;
+  }
+
+  if (has_validity) {
+    __syncthreads(); // wait for shared null counts to be ready
+
+    // Compute total null_count change for this block and add it to global count
+    if (threadIdx.x < warp_size) {
+      uint32_t my_null_change = warp_null_change[threadIdx.x];
+
+      __shared__ typename cub::WarpReduce<uint32_t>::TempStorage temp_storage;
+
+      uint32_t block_null_change =
+        cub::WarpReduce<uint32_t>(temp_storage).Sum(my_null_change);
+
+      if (lane_id == 0) { // one thread computes and adds to null count
+        atomicAdd(null_count, block_null_change);
+      }
+    }
+  }
+}
+
+template <typename InputFactory>
+struct copy_range_dispatch {
+  InputFactory factory;
+
+  template <typename T>
+  void operator()(cudf::mutable_column_view& output,
+                  cudf::size_type begin, cudf::size_type end,
+                  cudaStream_t stream = 0) {
+    static_assert(warp_size == cudf::util::size_in_bits<cudf::bitmask_type>(),
+      "copy_range_kernel assumes bitmask element size in bits == warp size");
+
+    auto input = factory.template make<T>();
+
+    T * __restrict__ data = output.head<T>();
+    cudf::bitmask_type * __restrict__ bitmask = output.null_mask();
+    auto offset = output.offset();
+
+#if 1
+    auto warp_aligned_begin_lower_bound =
+      cudf::size_type{begin - (begin % warp_size)};
+    auto warp_aligned_end_upper_bound =
+      cudf::size_type{
+        (end % warp_size) == 0 ? end : end + (warp_size - (end % warp_size))};
+    auto num_items =
+      warp_aligned_end_upper_bound - warp_aligned_begin_lower_bound;
+#else
+    // This one results in a compiler internal error! TODO: file NVIDIA bug
+    // cudf::size_type num_items = cudf::util::round_up_safe(end - begin, warp_size);
+    // number threads to cover range, rounded to nearest warp
+    // this code runs for one additional round if begin is not warp aligned,
+    // and end is block_size + 1
+    auto num_items =
+      cudf::size_type{
+        warp_size * cudf::util::div_rounding_up_safe(end - begin, warp_size)};
```

Comment:
Is this comment about the comments on the compilation error or the new code to compute the num_items?

That comments (Line 150 & 151) about the compilation error is from the legacy code and Line 155 - 157 compiles fine, but this leads the while loop (https://github.com/rapidsai/cudf/pull/3172/files/5b875f11844e7b7159b18032d366c64cca4c7fc6#diff-ce0f8bda3307b2967db2b1d5485ac9e9R76) in the copy_range_kernel function to loop one more times for certain values of begin and end (e.g. begin = 31 & end 257).

The copy range kernel works on warp granularities, so if begin = 31 and end = 257, it works from 0 to 288; i.e. [0, 288).

But the commented out code computes num_items as a function of just (end - begin = 226, does not consider that begin is not warp-aligned), and
```
auto grid = cudf::util::cuda::grid_config_1d{num_items, block_size, 1};
```
produces a grid of size 256 (which is smaller than 288), and this leads the above while loop to run two times instead of one. This code is to fix this, and please let me know if I am misunderstanding something.